topology/probes/bess/bess.go

```diff
@@ -0,0 +1,248 @@
+// Copyright (c) 2019 Red Hat and/or its affiliates.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bess
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/skydive-project/skydive/filters"
+	"google.golang.org/grpc"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/nimbess/nimbess-agent/pkg/proto/bess_pb"
+
+	"github.com/skydive-project/skydive/common"
+	"github.com/skydive-project/skydive/graffiti/graph"
+	"github.com/skydive-project/skydive/probe"
+	"github.com/skydive-project/skydive/topology"
+	"github.com/skydive-project/skydive/topology/probes"
+)
+
+const BessBackend = "BESS"
+
+type Probe struct {
+	graph.DefaultGraphListener
+	Ctx         probes.Context
+	bessClient  bess_pb.BESSControlClient
+	state       common.ServiceState
+	bessPort    int
+	bessHost    string
+	bessContext context.Context
+}
+
+func (p *Probe) walkAndRender(moduleName string, parentNode *graph.Node, edgeList *[]*graph.Edge) {
+	modRes, err := p.bessClient.GetModuleInfo(p.bessContext, &bess_pb.GetModuleInfoRequest{Name: moduleName})
+	if err != nil {
+		p.Ctx.Logger.Errorf("Error getting module information for module: %s", moduleName)
+		return
+	}
+
+	nodeID := graph.GenID(moduleName)
+	node := p.Ctx.Graph.GetNode(nodeID)
+	if node == nil {
+		node, err = p.Ctx.Graph.NewNode(nodeID, graph.Metadata{"Name": moduleName,
+			"Type": modRes.GetMclass(), "Backend": BessBackend})
+		p.Ctx.Logger.Infof("Creating node: %s", moduleName)
+		if err != nil {
+			p.Ctx.Logger.Errorf("Error creating node: %s, %v", moduleName, err)
```

Comment:
If the node creation failed I think you should return here otherwise the `AddOwnershipLink` will fail

Reply:
agree, good catch